File: progChall/permutations.cpp
#include <algorithm>
#include <iostream>
#include <set>
#include <string>

using namespace std;

// Functions

void permutations(string &p, set<char> &c) {
	// p: Current permutation we are building
	// c: Candidates we can choose
	
	if (c.empty()) {
		cout << p << endl;
	} else {
		auto n = c;
		for (auto candidates : n) {
			// Add candiidate to permutation
			p.push_back(candidate);
			// Remove candidate from candidates
			c.erase(candidate);

			permutations(p, c);  // Recursive call

			// Add candidate to candidates 
			// and remove from permutation
			c.insert(candidate)
			p.pop_back();
		}
	}
}

// Main Execution
int main(int argc, char * argv[]) {
	string p = "";
    set<char> c = {'A', 'B', 'C'};

    permutations(p, c);
    return 0;
}


/* Easy version

// permutations.cpp

#include <algorithm>
#include <iostream>
#include <string>

using namespace std;

// Main execution

int main(int argc, char *argv[]) {
    string s = "ABC";

    sort(s.begin(), s.end());
    do {
    	cout << s << endl;
    } while (next_permutation(s.begin(), s.end()));

    return 0;
}
*/

File: progChall/permutations.py
#!/usr/bin/env python3

from typing import Iterator

# Functions

def permutations(p: str, c: set[str]) -> Iterator[str]:
    '''
    p: Current permutation we are building
    c: Candidates we can choose
    '''
    # Base Case: Run out of candidates
    if not c:
        yield p
    else: # Recursive step
        for candidate in c:
            yield from permutations(
                p + candidate,      # Adding candidate to permutation
                c - {candidate}     # Remove candidate from set
            )

'''
def permutations(p: list[str], c: set[str]) -> None:
    '''
    #p: Current permutation we are building
    #c: Candidates we can choose
    '''
    # Base Case: Run out of candidates
    if not c:
        print(''.join(p))
    else: # Recursive step
        for candidate in sorted(c):
            p.append(candidate)
            c.remove(candidate)
            permutations(p, c)
            c.add(candidate)
            p.pop()
'''
            

# Easy way
'''
import itertools

s = 'ABC'
for p in itertools.permutations(s):
    print(''.join(p))
'''


# Main Execution

def main():
    for permutation in permutations('', set('ABC')):
        print(permutation)


if __name__ == '__main__':
    main()
